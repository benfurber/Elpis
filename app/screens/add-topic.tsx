import React, { Component } from "react";
import { StyleSheet, Text, TextInput, View } from "react-native";

import { Mutation } from "react-apollo";
import { withMappedNavigationParams } from "react-navigation-props-mapper";

import { NavigationType, Post } from "interfaces";
import { labels } from "labels";
import { ADD_COMMENT } from "mutations";
import { FormContainerScreen } from "screens";
import { layout, typography, colours } from "styles";
import { firstSentence } from "utils";

interface Props {
  navigation: NavigationType;
  postId: Post["id"];
}

interface State {
  editable: boolean;
  textInput: string;
  textInputExtra: string;
}

class AddTopicScreen extends Component<Props, State> {
  constructor(props) {
    super(props);
    this.state = {
      editable: true,
      textInput: "",
      textInputExtra: "",
    };
  }
  secondTextInput = TextInput as any;

  componentDidUpdate(_props, state) {
    const textInput = firstSentence(state.textInput);

    if (textInput !== state.textInput) {
      this.setState({ textInput });
      return this.secondTextInput.focus();
    }
  }

  onSubmitEditing(query, id) {
    const { textInput, textInputExtra } = this.state;
    this.setState({ editable: false });

    const content = textInput + " " + textInputExtra;

    query({
      variables: {
        content,
        id,
      },
    }).then(() => {
      this.setState({ textInput: "", editable: true });
      this.props.navigation.dismiss();
    });
  }

  form() {
    const { textInput, editable, textInputExtra } = this.state;

    const args = {
      editable,
      multiline: true,
      onChangeText: textInput => this.setState({ textInput }),
    };

    return (createComment, {}) => (
      <View style={styles.textInputContainer}>
        <Text style={styles.label}>{labels.title}</Text>
        <TextInput
          onSubmitEditing={() => this.secondTextInput.focus()}
          style={styles.title}
          autoFocus={true}
          placeholder={labels.addPlaceholderTitle}
          value={textInput}
          returnKeyLabel={"Next"}
          {...args}
        />
        <Text style={styles.label}>{labels.body}</Text>
        <TextInput
          {...args}
          autoFocus={false}
          ref={input => (this.secondTextInput = input)}
          onChangeText={textInputExtra => this.setState({ textInputExtra })}
          onSubmitEditing={() =>
            this.onSubmitEditing(createComment, this.props.postId)
          }
          value={textInputExtra}
          style={styles.text}
          placeholder={labels.addPlaceholderBody}
          returnKeyLabel={labels.submit}
        />
      </View>
    );
  }

  renderAddResponse(mutation) {
    return <Mutation mutation={mutation}>{this.form()}</Mutation>;
  }

  render() {
    return (
      <FormContainerScreen
        analyticsContent={{
          contentType: "AddTopic",
          contentId: this.props.postId,
        }}
        navigation={this.props.navigation}
        title={labels.addNewTopic}
      >
        <View style={styles.row}>{this.renderAddResponse(ADD_COMMENT)}</View>
      </FormContainerScreen>
    );
  }
}

const styles = StyleSheet.create({
  label: {
    color: colours.mediumGrey,
    fontWeight: "bold",
    paddingTop: layout.spacingXL,
    paddingHorizontal: layout.spacingL,
  },
  row: {
    alignItems: "baseline",
    flexDirection: "row",
  },
  textInputContainer: {
    width: "100%",
    flexDirection: "column",
  },
  title: {
    padding: layout.spacingL,
    fontFamily: "creteround-regular",
    fontSize: typography.fontSizeXL,
    flexWrap: "wrap",
  },
  text: {
    fontFamily: "creteround-regular",
    fontSize: typography.fontSizeL,
    padding: layout.spacingL,
    width: "100%",
  },
});

const wrappedAddTopicScreen = withMappedNavigationParams()(AddTopicScreen);
export {
  wrappedAddTopicScreen as AddTopicScreen,
  AddTopicScreen as UnwrappedAddTopicScreen,
};
