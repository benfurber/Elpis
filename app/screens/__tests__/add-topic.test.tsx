import "react-native";
import React from "react";
import { shallow } from "enzyme";

import { UnwrappedAddTopicScreen } from "../add-topic";

let navigation;
jest.mock(navigation, () => jest.fn());

describe("AddReplyScreen", () => {
  it("renders correctly", () => {
    const component = shallow(
      <UnwrappedAddTopicScreen navigation={navigation} postId={"2332r"} />,
    );

    expect(component).toMatchSnapshot();
  });
});
