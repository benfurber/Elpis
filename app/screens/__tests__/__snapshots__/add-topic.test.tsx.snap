// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddReplyScreen renders correctly 1`] = `
<FormContainerScreen
  analyticsContent={
    Object {
      "contentId": "2332r",
      "contentType": "AddTopic",
    }
  }
  title="Adicione um novo tópico na discussão"
>
  <View
    style={
      Object {
        "alignItems": "baseline",
        "flexDirection": "row",
      }
    }
  >
    <Mutation
      mutation={
        Object {
          "definitions": Array [
            Object {
              "directives": Array [],
              "kind": "OperationDefinition",
              "name": Object {
                "kind": "Name",
                "value": "createComment",
              },
              "operation": "mutation",
              "selectionSet": Object {
                "kind": "SelectionSet",
                "selections": Array [
                  Object {
                    "alias": undefined,
                    "arguments": Array [
                      Object {
                        "kind": "Argument",
                        "name": Object {
                          "kind": "Name",
                          "value": "content",
                        },
                        "value": Object {
                          "kind": "Variable",
                          "name": Object {
                            "kind": "Name",
                            "value": "content",
                          },
                        },
                      },
                      Object {
                        "kind": "Argument",
                        "name": Object {
                          "kind": "Name",
                          "value": "id",
                        },
                        "value": Object {
                          "kind": "Variable",
                          "name": Object {
                            "kind": "Name",
                            "value": "id",
                          },
                        },
                      },
                    ],
                    "directives": Array [],
                    "kind": "Field",
                    "name": Object {
                      "kind": "Name",
                      "value": "createComment",
                    },
                    "selectionSet": Object {
                      "kind": "SelectionSet",
                      "selections": Array [
                        Object {
                          "alias": undefined,
                          "arguments": Array [],
                          "directives": Array [],
                          "kind": "Field",
                          "name": Object {
                            "kind": "Name",
                            "value": "id",
                          },
                          "selectionSet": undefined,
                        },
                      ],
                    },
                  },
                ],
              },
              "variableDefinitions": Array [
                Object {
                  "defaultValue": undefined,
                  "directives": Array [],
                  "kind": "VariableDefinition",
                  "type": Object {
                    "kind": "NonNullType",
                    "type": Object {
                      "kind": "NamedType",
                      "name": Object {
                        "kind": "Name",
                        "value": "String",
                      },
                    },
                  },
                  "variable": Object {
                    "kind": "Variable",
                    "name": Object {
                      "kind": "Name",
                      "value": "content",
                    },
                  },
                },
                Object {
                  "defaultValue": undefined,
                  "directives": Array [],
                  "kind": "VariableDefinition",
                  "type": Object {
                    "kind": "NonNullType",
                    "type": Object {
                      "kind": "NamedType",
                      "name": Object {
                        "kind": "Name",
                        "value": "ID",
                      },
                    },
                  },
                  "variable": Object {
                    "kind": "Variable",
                    "name": Object {
                      "kind": "Name",
                      "value": "id",
                    },
                  },
                },
              ],
            },
          ],
          "kind": "Document",
          "loc": Object {
            "end": 124,
            "start": 0,
          },
        }
      }
    >
      <Component />
    </Mutation>
  </View>
</FormContainerScreen>
`;
